src/OrbitGl/GraphTrack.cpp

```diff
@@ -0,0 +1,293 @@
+  std::string text = GetLabelTextFromValues(values);
+  DrawLabel(batcher, text_renderer, Vec2(point_x, point_y), text, kBlack, kWhite, label_z);
+
+  // Draw legends
+  if (series_.names().size() <= 1 || collapse_toggle_->IsCollapsed()) return;
+  DrawLegend(batcher, text_renderer, series_.names(), kWhite, text_z);
+}
+
+template <size_t Size>
+void GraphTrack<Size>::UpdatePrimitives(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                        PickingMode picking_mode, float z_offset) {
+  float track_width = viewport_->GetVisibleWorldWidth();
+  SetSize(track_width, GetHeight());
+  pos_[0] = viewport_->GetWorldTopLeft()[0];
+
+  float track_z = GlCanvas::kZValueTrack + z_offset;
+  float graph_z = GlCanvas::kZValueEventBar + z_offset;
+
+  float content_height =
+      size_[1] - layout_->GetTrackTabHeight() - layout_->GetTrackBottomMargin() - GetLegendHeight();
+  Vec2 content_pos = pos_;
+  content_pos[1] -= layout_->GetTrackTabHeight();
+  Box box(content_pos, Vec2(size_[0], -content_height - GetLegendHeight()), track_z);
+  batcher->AddBox(box, GetTrackBackgroundColor(), shared_from_this());
+
+  const bool picking = picking_mode != PickingMode::kNone;
+  if (!picking) {
+    double time_range = static_cast<double>(max_tick - min_tick);
+    if (series_.value_map().size() < 2 || time_range == 0) return;
+    DrawSeries(batcher, min_tick, max_tick, graph_z);
+  }
+}
+
+template <size_t Size>
+void GraphTrack<Size>::SetLabelUnitWhenEmpty(const std::string& label_unit) {
+  if (!label_unit_.empty()) return;
+  label_unit_ = label_unit;
+}
+
+template <size_t Size>
+void GraphTrack<Size>::SetValueDecimalDigitsWhenEmpty(uint8_t value_decimal_digits) {
+  if (value_decimal_digits_.has_value()) return;
+  value_decimal_digits_ = value_decimal_digits;
+}
+
+template <size_t Size>
+void GraphTrack<Size>::OnTimer(const orbit_client_protos::TimerInfo& timer_info) {
+  constexpr uint32_t kDepth = 0;
+  std::shared_ptr<TimerChain> timer_chain = timers_[kDepth];
+  if (timer_chain == nullptr) {
+    timer_chain = std::make_shared<TimerChain>();
+    timers_[kDepth] = timer_chain;
+  }
+
+  timer_chain->emplace_back(timer_info);
+}
+
+template <size_t Size>
+std::vector<std::shared_ptr<TimerChain>> GraphTrack<Size>::GetAllChains() const {
+  std::vector<std::shared_ptr<TimerChain>> chains;
+  for (const auto& pair : timers_) {
+    chains.push_back(pair.second);
+  }
+  return chains;
+}
+
+template <size_t Size>
+std::optional<std::pair<uint64_t, std::array<double, Size>>>
+GraphTrack<Size>::GetPreviousTimeAndValues(uint64_t time) const {
+  std::map<uint64_t, std::array<double, Size>> time_to_series_values = series_.value_map();
+  auto iterator_lower = time_to_series_values.upper_bound(time);
+  if (iterator_lower == time_to_series_values.begin()) return {};
+  return *(--iterator_lower);
+}
+
+template <size_t Size>
+float GraphTrack<Size>::GetLabelYFromValues(const std::array<double, Size>& /*values*/) const {
```

Comment:
This is because the label Y location might depend on the values. More specifically:
* For the `VariableTrack`: let `y` be the Y location of the line. `y` is determined by the value and white label's Y location equals to `y`;
* For the `MemoryUsageTrack`:  let `y` be the Y location of the top of rectangles. `y` is determined by the sum of values, which always equals to the total memory usage. And the white label's Y location equals `y / 2.f`. 
* For the `PagefaultTrack`: let `y` be the Y location of the most above line. `y` is determined by the maximum of values and the white label's Y location can be `y / 2.f`.

Reply:
Thanks!